File: frontend/src/api/clusterAnalysis.ts
/**
 * Cluster Analysis API - 统一的 Cluster 高级计算规划
 */
import apiClient from './client';

// ============================================================================
// 类型定义
// ============================================================================

export type ClusterCalcType =
  | 'BINDING_TOTAL'
  | 'BINDING_PAIRWISE'
  | 'DESOLVATION_STEPWISE'
  | 'DESOLVATION_FULL'
  | 'REDOX'
  | 'REORGANIZATION';

export type AdvancedClusterJobStatus =
  | 'CREATED'
  | 'SUBMITTED'
  | 'RUNNING'
  | 'WAITING_QC'
  | 'CALCULATING'
  | 'COMPLETED'
  | 'FAILED'
  | 'CANCELLED';

export interface QCConfig {
  functional: string;
  basis_set: string;
  solvent_model?: string;
  solvent?: string;
  use_dispersion: boolean;
  charge_cluster: number;
  charge_ion: number;
  // Slurm 资源配置
  slurm_partition?: string;
  slurm_cpus?: number;
  slurm_time?: number;
}

export interface PlannedQCTask {
  task_type: string;
  description: string;
  smiles?: string;
  structure_id?: number;
  charge: number;
  multiplicity: number;
  calc_mode: 'opt' | 'sp';  // 计算模式: opt (几何优化) / sp (单点能量)
  status: 'new' | 'reused' | 'local_reused' | 'pending';
  existing_qc_job_id?: number;
  existing_energy?: number;
}

export interface CalcTypeRequirements {
  calc_type: ClusterCalcType;
  description: string;
  required_qc_tasks: PlannedQCTask[];
  new_tasks_count: number;
  reused_tasks_count: number;
}

export interface RedoxOptions {
  include_molecule: boolean;
  include_dimer: boolean;
  include_cluster: boolean;
}

export interface ReorganizationOptions {
  include_molecule: boolean;
  include_cluster: boolean;
}

export interface ClusterAnalysisPlanRequest {
  md_job_id: number;
  solvation_structure_ids?: number[];
  composition_keys?: string[];
  calc_types: ClusterCalcType[];
  redox_options?: RedoxOptions;
  reorganization_options?: ReorganizationOptions;
  qc_config?: Partial<QCConfig>;
}

export interface ClusterAnalysisPlanResponse {
  md_job_id: number;
  selected_structures_count: number;
  selected_structure_ids: number[];
  calc_requirements: CalcTypeRequirements[];
  total_new_qc_tasks: number;
  total_reused_qc_tasks: number;
  estimated_compute_hours: number;
  warnings: string[];
}

export interface AdvancedClusterJob {
  id: number;
  md_job_id: number;
  user_id: number;
  username?: string;  // 仅 admin 可见
  user_email?: string;  // 仅 admin 可见
  status: AdvancedClusterJobStatus;
  progress: number;
  calc_types: string[];
  selected_structures: {
    solvation_structure_ids: number[];
    count: number;
  };
  qc_config: Record<string, unknown>;
  qc_task_plan: {
    planned_qc_tasks: PlannedQCTask[];
    reused_qc_jobs: number[];
    new_qc_jobs: number[];
    total_qc_tasks: number;
    completed_qc_tasks: number;
  };
  results: Record<string, unknown>;
  error_message?: string;
  cpu_hours_used?: number;  // 实际消耗的核时
  task_count?: number;  // 任务计数
  created_at: string;
  updated_at: string;
  started_at?: string;
  finished_at?: string;
}

export interface AddCalcTypePlanResponse {
  job_id: number;
  existing_calc_types: string[];
  additional_calc_types: string[];
  new_qc_tasks_required: number;
  reused_from_existing: number;
  details: CalcTypeRequirements[];
}

// ============================================================================
// 计算类型描述
// ============================================================================

export const CALC_TYPE_INFO: Record<ClusterCalcType, {
  label: string;
  description: string;
  formula: string;
  riskLevel: 'low' | 'medium' | 'high';
  icon: string;
}> = {
  BINDING_TOTAL: {
    label: '总 Binding Energy',
    description: '计算整个溶剂化簇的总脱溶剂化能',
    formula: 'E_bind = E_cluster - (E_ion + Σ n_j × E_ligand_j)',
    riskLevel: 'low',
    icon: '🔗',
  },
  BINDING_PAIRWISE: {
    label: '分子-离子 Binding',
    description: '计算单个分子与中心离子的 binding energy',
    formula: 'E_bind = E(Ion-X) - E(Ion) - E(X)',
    riskLevel: 'low',
    icon: '⚛️',
  },
  DESOLVATION_STEPWISE: {
    label: '逐级去溶剂化',
    description: '逐个移除配体，计算每步的去溶剂化能',
    formula: 'ΔE_i = E_cluster - (E_minus_i + E_ligand_i)',
    riskLevel: 'medium',
    icon: '📉',
  },
  DESOLVATION_FULL: {
    label: '完全去溶剂化',
    description: '计算从完整簇到裸离子的总去溶剂化能',
    formula: 'ΔE = E_cluster - (E_ion + Σ E_ligand_i)',
    riskLevel: 'low',
    icon: '🎯',
  },
  REDOX: {
    label: '氧化还原电位',
    description: '热力学循环计算氧化还原电位',
    formula: 'E° = -ΔG(sol) / nF',
    riskLevel: 'high',
    icon: '⚡',
  },
  REORGANIZATION: {
    label: 'Marcus 重组能',
    description: 'Marcus 理论 4 点方案计算重组能',
    formula: 'λ = (λ_ox + λ_red) / 2',
    riskLevel: 'high',
    icon: '🔄',
  },
};

// ============================================================================
// API 函数
// ============================================================================

export interface RecommendedSolventResponse {
  recommended_solvent: string;
  recommended_dielectric: number;
  average_dielectric: number;
  composition_analyzed: Record<string, { count: number; dielectric: number }>;
  reason: string;
}

export async function recommendPCMSolvent(
  mdJobId: number
): Promise<RecommendedSolventResponse> {
  const response = await apiClient.get(`/cluster-analysis/recommend-solvent/${mdJobId}`);
  return response.data;
}

export async function planClusterAnalysis(
  request: ClusterAnalysisPlanRequest
): Promise<ClusterAnalysisPlanResponse> {
  const response = await apiClient.post('/cluster-analysis/plan', request, {
    timeout: 90000 // 90秒超时，因为规划可能需要较长时间
  });
  return response.data;
}

export async function submitClusterAnalysis(
  request: ClusterAnalysisPlanRequest
): Promise<AdvancedClusterJob> {
  const response = await apiClient.post('/cluster-analysis/submit', request, {
    timeout: 120000 // 2分钟超时，因为提交可能需要创建大量QC任务
  });
  return response.data;
}

export async function listClusterAnalysisJobs(
  mdJobId?: number,
  skip = 0,
  limit = 50
): Promise<AdvancedClusterJob[]> {
  const params: Record<string, unknown> = { skip, limit };
  if (mdJobId) params.md_job_id = mdJobId;
  const response = await apiClient.get('/cluster-analysis/jobs', { params });
  return response.data;
}

export async function getClusterAnalysisJob(jobId: number): Promise<AdvancedClusterJob> {
  const response = await apiClient.get(`/cluster-analysis/jobs/${jobId}`);
  return response.data;
}

export async function cancelClusterAnalysisJob(jobId: number): Promise<{
  message: string;
  job_id: number;
  cancelled_qc_count: number;
}> {
  const response = await apiClient.post(`/cluster-analysis/jobs/${jobId}/cancel`);
  return response.data;
}

export async function resubmitClusterAnalysisJob(jobId: number): Promise<AdvancedClusterJob> {
  const response = await apiClient.post(`/cluster-analysis/jobs/${jobId}/resubmit`);
  return response.data;
}

export async function deleteClusterAnalysisJob(jobId: number): Promise<{
  message: string;
  job_id: number;
  action: string;
}> {
  const response = await apiClient.delete(`/cluster-analysis/jobs/${jobId}`);
  return response.data;
}

export async function planAddCalcTypes(
  jobId: number,
  additionalCalcTypes: ClusterCalcType[]
): Promise<AddCalcTypePlanResponse> {
  const response = await apiClient.post(`/cluster-analysis/jobs/${jobId}/add-calc-types`, {
    job_id: jobId,
    additional_calc_types: additionalCalcTypes,
  });
  return response.data;
}

// ============================================================================
// 结果查询 API
// ============================================================================

export interface QCTaskInfo {
  task_type: string;
  name: string;
  description: string;
  smiles: string;
  charge: number;
  multiplicity: number;
  status: 'new' | 'reused' | 'local_reused';
  qc_job_id: number | null;
  qc_status: string | null;
  slurm_job_id?: string | null;
  functional?: string;
  basis_set?: string;
  solvent_model?: string;
  solvent_name?: string;
}

export interface QCStatus {
  job_id: number;
  total_qc_jobs: number;
  completed: number;
  running: number;
  pending: number;
  failed: number;
  all_completed: boolean;
  calc_types?: string[];
  tasks_by_calc_type?: Record<string, QCTaskInfo[]>;
  // CPU 核时统计（真实的 Slurm CPUTimeRAW）
  total_cpu_hours?: number;
  completed_cpu_hours?: number;
  running_cpu_hours?: number;
  qc_jobs?: Array<{
    id: number;
    status: string;
    molecule_name: string;
    task_type?: string;
    actual_cpu_hours?: number;
  }>;
}

export interface ClusterAnalysisResults {
  job_id: number;
  status: string;
  progress: number;
  calc_types: string[];
  results: Record<string, unknown>;
  qc_task_plan: Record<string, unknown>;
}

// Binding Total 结果
export interface BindingTotalResult {
  e_cluster_au: number;
  e_ion_au: number;
  ligand_energies_au: Record<string, number>;
  e_bind_au: number;
  e_bind_ev: number;
  e_bind_kcal_mol: number;
  error?: string;
}

// Binding Pairwise 结果
export interface BindingPairwiseResult {
  pairwise_bindings: Array<{
    ligand: string;
    e_dimer_au: number;
    e_ligand_au: number;
    e_ion_au: number;
    e_bind_au: number;
    e_bind_ev: number;
    e_bind_kcal_mol: number;
  }>;
}

// Desolvation Stepwise 结果
export interface DesolvationStepwiseResult {
  stepwise_desolvation: Array<{
    ligand: string;
    e_cluster_au: number;
    e_minus_au: number;
    e_ligand_au: number;
    delta_e_au: number;
    delta_e_ev: number;
    delta_e_kcal_mol: number;
  }>;
}

// Redox 结果
export interface RedoxResult {
  redox_potentials: Array<{
    smiles: string;
    e_neutral_gas_au: number;
    e_charged_gas_au: number;
    e_neutral_sol_au: number;
    e_charged_sol_au: number;
    delta_g_sol_ev: number;
    oxidation_potential_v: number;
  }>;
}

export async function getClusterAnalysisResults(jobId: number): Promise<ClusterAnalysisResults> {
  const response = await apiClient.get(`/cluster-analysis/jobs/${jobId}/results`);
  return response.data;
}

export async function getClusterAnalysisQCStatus(jobId: number): Promise<QCStatus> {
  const response = await apiClient.get(`/cluster-analysis/jobs/${jobId}/qc-status`);
  return response.data;
}

export async function calculateClusterAnalysisResults(jobId: number): Promise<{ status: string; results: Record<string, unknown> }> {
  const response = await apiClient.post(`/cluster-analysis/jobs/${jobId}/calculate`);
  return response.data;
}

File: backend/app/schemas/cluster_analysis.py
"""
Cluster Analysis Schemas - 统一的 Cluster 高级计算规划
"""
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from enum import Enum
from datetime import datetime


class ClusterCalcType(str, Enum):
    """Cluster 计算类型"""
    BINDING_TOTAL = "BINDING_TOTAL"           # Binding A: 总脱溶剂化能
    BINDING_PAIRWISE = "BINDING_PAIRWISE"     # Binding B: 单分子-Li Binding
    DESOLVATION_STEPWISE = "DESOLVATION_STEPWISE"  # 逐级去溶剂化能
    DESOLVATION_FULL = "DESOLVATION_FULL"     # 完全去溶剂化能
    REDOX = "REDOX"                           # 氧化还原电位
    REORGANIZATION = "REORGANIZATION"         # Marcus 重组能


class AdvancedClusterJobStatus(str, Enum):
    """高级 Cluster 计算任务状态"""
    CREATED = "CREATED"
    SUBMITTED = "SUBMITTED"
    RUNNING = "RUNNING"
    WAITING_QC = "WAITING_QC"
    CALCULATING = "CALCULATING"
    COMPLETED = "COMPLETED"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


# ============================================================================
# 规划请求/响应
# ============================================================================

class QCConfig(BaseModel):
    """QC 计算配置"""
    functional: str = Field("B3LYP", description="DFT 泛函")
    basis_set: str = Field("6-31G*", description="基组")
    solvent_model: Optional[str] = Field(None, description="溶剂模型 (SMD/PCM/None)")
    solvent: Optional[str] = Field(None, description="溶剂名称")
    use_dispersion: bool = Field(True, description="是否使用色散校正 (D3BJ)")
    charge_cluster: int = Field(1, description="Cluster 电荷")
    charge_ion: int = Field(1, description="离子电荷 (Li+ = 1)")
    # Slurm 资源配置
    slurm_partition: Optional[str] = Field("cpu", description="Slurm 队列/分区")
    slurm_cpus: Optional[int] = Field(16, ge=1, le=128, description="CPU 核心数")
    slurm_time: Optional[int] = Field(7200, ge=10, description="最大运行时间（分钟）")


class RedoxOptions(BaseModel):
    """REDOX 子选项"""
    include_molecule: bool = Field(True, description="计算单独配体分子的氧化还原电位")
    include_dimer: bool = Field(True, description="计算 Li-配体 Dimer 的氧化还原电位")
    include_cluster: bool = Field(False, description="计算整个 Cluster 的氧化还原电位（计算量大）")


class ReorganizationOptions(BaseModel):
    """REORGANIZATION 子选项"""
    include_molecule: bool = Field(True, description="计算单独配体分子的重组能")
    include_cluster: bool = Field(True, description="计算整个 Cluster 的重组能")


class ClusterAnalysisPlanRequest(BaseModel):
    """规划请求 - 用户选择结构和计算类型"""
    md_job_id: int = Field(..., description="MD 任务 ID")

    # 选中的结构（二选一）
    solvation_structure_ids: Optional[List[int]] = Field(None, description="选中的溶剂化结构 ID 列表")
    composition_keys: Optional[List[str]] = Field(None, description="选中的 composition_key 列表")

    # 选中的计算类型
    calc_types: List[ClusterCalcType] = Field(..., description="选中的计算类型")

    # 子选项
    redox_options: Optional[RedoxOptions] = Field(default_factory=RedoxOptions, description="REDOX 子选项")
    reorganization_options: Optional[ReorganizationOptions] = Field(default_factory=ReorganizationOptions, description="REORGANIZATION 子选项")

    # QC 配置
    qc_config: Optional[QCConfig] = Field(default_factory=QCConfig, description="QC 计算配置")


class PlannedQCTask(BaseModel):
    """规划的 QC 任务"""
    task_type: str = Field(..., description="任务类型: cluster/ligand/ion/cluster_minus/dimer/charged")
    description: str = Field(..., description="任务描述")
    smiles: Optional[str] = Field(None, description="分子 SMILES")
    structure_id: Optional[int] = Field(None, description="关联的结构 ID")
    charge: int = Field(0, description="电荷")
    multiplicity: int = Field(1, description="多重度")
    calc_mode: str = Field("opt", description="计算模式: opt (几何优化) / sp (单点能量)")

    # 复用状态
    status: str = Field("new", description="状态: new/reused/pending")
    existing_qc_job_id: Optional[int] = Field(None, description="已有的 QC 任务 ID（复用时）")
    existing_energy: Optional[float] = Field(None, description="已有的能量值（Hartree）")


class CalcTypeRequirements(BaseModel):
    """某个计算类型的 QC 需求"""
    calc_type: ClusterCalcType
    description: str
    required_qc_tasks: List[PlannedQCTask]
    new_tasks_count: int = Field(0, description="需要新建的 QC 任务数")
    reused_tasks_count: int = Field(0, description="可复用的 QC 任务数")


class ClusterAnalysisPlanResponse(BaseModel):
    """规划响应 - 返回 QC 任务规划详情"""
    md_job_id: int
    
    # 选中的结构信息
    selected_structures_count: int
    selected_structure_ids: List[int]
    
    # 各计算类型的 QC 需求
    calc_requirements: List[CalcTypeRequirements]
    
    # 汇总
    total_new_qc_tasks: int = Field(..., description="总共需要新建的 QC 任务数")
    total_reused_qc_tasks: int = Field(..., description="总共可复用的 QC 任务数")
    estimated_compute_hours: float = Field(0.0, description="预估计算时间（小时）")
    
    # 警告
    warnings: List[str] = Field(default_factory=list, description="警告信息")


# ============================================================================
# 提交/创建任务
# ============================================================================

class ClusterAnalysisSubmitRequest(BaseModel):
    """提交计算任务请求"""
    md_job_id: int = Field(..., description="MD 任务 ID")
    solvation_structure_ids: Optional[List[int]] = Field(None, description="选中的溶剂化结构 ID 列表")
    composition_keys: Optional[List[str]] = Field(None, description="选中的 composition_key 列表")
    calc_types: List[ClusterCalcType] = Field(..., description="选中的计算类型")
    qc_config: Optional[QCConfig] = Field(default_factory=QCConfig, description="QC 计算配置")
    # 子选项（与 PlanRequest 保持一致）
    redox_options: Optional[RedoxOptions] = Field(default_factory=RedoxOptions, description="REDOX 子选项")
    reorganization_options: Optional[ReorganizationOptions] = Field(default_factory=ReorganizationOptions, description="REORGANIZATION 子选项")


class AdvancedClusterJobResponse(BaseModel):
    """任务响应"""
    id: int
    md_job_id: int
    user_id: int
    username: Optional[str] = None  # 仅 admin 可见
    user_email: Optional[str] = None  # 仅 admin 可见
    status: AdvancedClusterJobStatus
    progress: float
    calc_types: List[str]
    selected_structures: Dict[str, Any]
    qc_config: Dict[str, Any]
    qc_task_plan: Dict[str, Any]
    results: Dict[str, Any]
    error_message: Optional[str]
    cpu_hours_used: float = 0.0  # 实际消耗的核时
    task_count: int = 0  # 任务计数
    created_at: datetime
    updated_at: datetime
    started_at: Optional[datetime]
    finished_at: Optional[datetime]

    class Config:
        from_attributes = True


# ============================================================================
# 追加计算
# ============================================================================

class AddCalcTypeRequest(BaseModel):
    """追加计算类型请求"""
    job_id: int = Field(..., description="已有任务 ID")
    additional_calc_types: List[ClusterCalcType] = Field(..., description="追加的计算类型")


class AddCalcTypePlanResponse(BaseModel):
    """追加计算规划响应"""
    job_id: int
    existing_calc_types: List[str]
    additional_calc_types: List[str]
    new_qc_tasks_required: int
    reused_from_existing: int
    details: List[CalcTypeRequirements]

